taichi/backends/cuda/cuda_context.cpp

```diff
@@ -84,7 +84,7 @@ void CUDAContext::launch(void *func,
 
   KernelProfilerBase::TaskHandle task_handle;
   // Kernel launch
-  if (profiler_) {
+  if (profiler_ && profiler_->is_enabled()) {
```

Comment:
Here might need a bit clarification: when profiler_ is not null, i assume it's should be enabled? and vice versa?

Reply:
It seems that from the beginning make_profiler has been designed to return valid pointers
[taichi/kernel_profiler.cpp at v0.7.10 ](https://github.com/taichi-dev/taichi/blob/v0.7.10/taichi/program/kernel_profiler.cpp#L174) vs [taichi/kernel_profiler.cpp at master](https://github.com/taichi-dev/taichi/blob/master/taichi/program/kernel_profiler.cpp#L116)
[taichi/program.cpp at v0.7.10](https://github.com/taichi-dev/taichi/blob/v0.7.10/taichi/program/program.cpp#L123)  vs  [taichi/program.cpp at master](https://github.com/taichi-dev/taichi/blob/master/taichi/program/program.cpp#L71)

For CUDA backends, the difference is whether or not the profiler_ in the CUDAContext is set.
If kernel_profiler is not set, it will not be executed when kernels are launched.
[taichi/program.cpp at v0.7.10](https://github.com/taichi-dev/taichi/blob/v0.7.10/taichi/program/program.cpp#L130)  vs  [taichi/llvm_program.cpp at master](https://github.com/taichi-dev/taichi/blob/master/taichi/llvm/llvm_program.cpp#L94)

I tried to return `nullptr` but test on m1 did not pass [taichi-dev/taichi@f278285](https://github.com/taichi-dev/taichi/actions/runs/1456235456)

So I added an additional conditional statement to prevent external modifications from causing the profiler to be called accidentally

Reply:
Hmm but this check complicates the logic here, @yolo2themoon do we know why m1 failed with returning nullptr?

Reply:
removed the enable-related code, thanks to @ailzhang 's help in debugging